openfisca_uk/tools/takeup.py

```diff
@@ -0,0 +1,229 @@
+from typing import Tuple
+import numpy as np
+from openfisca_core.parameters import Parameter
+from openfisca_core.periods import instant
+from path import Path
+from tqdm import tqdm
+from openfisca_uk.system import CountryTaxBenefitSystem
+from openfisca_uk.tools.simulation import Microsimulation
+import logging
+import yaml
+
+logging.basicConfig(level=logging.INFO)
+
+
+class Program:
+    """A government program with receipients who must choose to apply."""
+
+    def __init__(
+        self,
+        variable: str,
+        caseload: Parameter,
+        expenditure: Parameter,
+        takeup: Parameter,
+        expenditure_weight: float = 0.5,
+    ):
+        self.variable = variable
+        self.caseload = caseload
+        self.caseload_name = caseload.name
+        self.expenditure = expenditure
+        self.expenditure_name = expenditure.name
+        self.takeup = takeup
+        self.takeup_name = takeup.name
+        self.expenditure_weight = expenditure_weight
+        self._reset_simulation(year=2018)
+
+    def _get_param(self, name: str):
+        param = self.microsimulation.simulation.tax_benefit_system.parameters
+        for part in name.split("."):
+            param = getattr(param, part)
+        return param
+
+    def _reset_simulation(self, year: int = 2018):
+        self.microsimulation = Microsimulation(
+            year=2018 if year == 2018 else 2019
+        )
+        self.variable_label = (
+            self.microsimulation.simulation.tax_benefit_system.variables[
+                self.variable
+            ].label
+        )
+        self.takeup = self._get_param(self.takeup_name)
+        self.caseload = self._get_param(self.caseload_name)
+        self.expenditure = self._get_param(self.expenditure_name)
+
+    def takeup_rate_error(self, takeup_rate: float, year: int) -> float:
+        if takeup_rate < 0 or takeup_rate > 1:
```

Comment:
Is this total takeup or takeup among households that are eligible but don't report claiming it? Total takeup can exceed 1.

Reply:
Just those that don't report it - will update the descriptions

Reply:
But although total takeup can exceed 1, we can't model it, as it just informs a probability